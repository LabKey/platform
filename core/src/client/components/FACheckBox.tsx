import React, { PureComponent } from 'react';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheckSquare } from '@fortawesome/free-solid-svg-icons';
import { faSquare } from '@fortawesome/free-regular-svg-icons';

interface Props {
    checked: boolean;
    canEdit: boolean;
    onClick: Function;
    name?: string;
}

export default class FACheckBox extends PureComponent<Props> {
    render() {
        const checkedOrNot = this.props.checked ? (
            <FontAwesomeIcon size="lg" icon={faCheckSquare} color="#0073BB" />
        ) : (
            <FontAwesomeIcon size="lg" icon={faSquare} color="#adadad" />
        );

        const classNames = this.props.canEdit ? 'no-highlight clickable ' : 'no-highlight ';

        return (
            <span className={classNames + (this.props.name ? this.props.name : "")} onClick={() => this.props.onClick()}>
                {checkedOrNot}
            </span>
        );
    }
}
