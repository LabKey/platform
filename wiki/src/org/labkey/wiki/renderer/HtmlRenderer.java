/*
 * Copyright (c) 2008-2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.labkey.wiki.renderer;

import org.apache.commons.lang3.StringUtils;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.labkey.api.attachments.Attachment;
import org.labkey.api.util.PageFlowUtil;
import org.labkey.api.util.TidyUtil;
import org.labkey.api.view.template.ClientDependency;
import org.labkey.api.wiki.WikiRenderer;
import org.labkey.api.wiki.FormattedHtml;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * User: Tamra Myers
 * Date: Aug 16, 2006
 * Time: 12:33:37 PM
 */
public class HtmlRenderer implements WikiRenderer
{
    private final String _hrefPrefix;
    private final String _attachPrefix;
    private final Map<String, String> _nameTitleMap;
    private final Map<String, Attachment> _attachments;

    private static Map<String, SubstitutionHandler> _substitutionHandlers = new HashMap<>();

    static
    {
        // Add SubstitutionHandlers for each ${labkey.<type>()}
        _substitutionHandlers.put("webPart", new WebPartSubstitutionHandler());
        _substitutionHandlers.put("dependency", new ClientDependencySubstitutionHandler());
    }


    public HtmlRenderer(String hrefPrefix, String attachPrefix, Map<String, String> nameTitleMap, @Nullable Collection<? extends Attachment> attachments)
    {
        _hrefPrefix = hrefPrefix;
        _attachPrefix = attachPrefix;
        _nameTitleMap = nameTitleMap == null ? new HashMap<>() : nameTitleMap;
        _attachments = new HashMap<>();

        if (null != attachments)
            for (Attachment a : attachments)
                _attachments.put(a.getName(), a);
    }
    

    public FormattedHtml format(String text)
    {
        LinkedList<String> errors = new LinkedList<>();
        if (text == null)
            return new FormattedHtml("");

        // Remove degenerate comments (e.g. "<!-->") as Tidy does not handle them properly
        text = text.replaceAll("<!--*>|<!>", "");

        FormattedHtml formattedHtml = handleLabKeySubstitutions(text);
        boolean volatilePage = formattedHtml.isVolatile();
        LinkedHashSet<ClientDependency> cds = new LinkedHashSet<>(formattedHtml.getClientDependencies());

        Document doc = TidyUtil.convertHtmlToDocument("<html><body>" + StringUtils.trimToEmpty(formattedHtml.getHtml()) + "</body></html>", false, errors);
        if (!errors.isEmpty() || doc == null)
        {
            StringBuilder innerHtml = new StringBuilder("<div class=\"labkey-error\"><b>An exception occurred while generating the HTML.  Please correct this content.</b></div><br>The error message was: ");
            if (!errors.isEmpty())
            {
                for (String error : errors)
                    innerHtml.append(PageFlowUtil.filter(error)).append("<hr>");
            }
            return new FormattedHtml(innerHtml.toString(), false, cds);
        }

        Set<String> wikiDependencies = new HashSet<>();

        // process A and IMG
        NodeList nl = doc.getElementsByTagName("a");
        Map<Element, String> linkExceptions = new HashMap<>();
        for (int i=0 ; i<nl.getLength() ; i++)
        {
            Element a = (Element)nl.item(i);
            try
            {
                String href = PageFlowUtil.decode(a.getAttribute("href"));

                Attachment at = _attachments.get(href);
                if (null != at)
                {
                    a.setAttribute("href", _attachPrefix + PageFlowUtil.encode(at.getName()));
                    wikiDependencies.add(at.getName());
                    continue;
                }

                if (href.isBlank() || href.startsWith("#"))
                    continue;

                String wikiName = StringUtils.substringBefore(href, "#");
                String title = _nameTitleMap.get(wikiName);
                if (null != title)
                {
                    // UNDONE: why is l.getName() null???
                    //a.setAttribute("href", _hrefPrefix + PageFlowUtil.encode(.getName()));
                    a.setAttribute("href", _hrefPrefix + PageFlowUtil.encode(href));
                    wikiDependencies.add(wikiName);
                    continue;
                }

                // if this doesn't look like a url, then link to it as a wiki page
                if (StringUtils.containsNone(href, "?:/.&"))
                {
                    a.setAttribute("href", _hrefPrefix + PageFlowUtil.encode(href));
                    wikiDependencies.add(wikiName);
                }
            }
            catch (IllegalArgumentException e)
            {
                // Issue 12160: there was an issue with decoding the href for the link, so store the exception text and link index for later
                linkExceptions.put(a, e.getMessage());
            }
        }

        nl = doc.getElementsByTagName("img");
        for (int i=0 ; i<nl.getLength() ; i++)
        {
            Element img = (Element)nl.item(i);
            String src = img.getAttribute("src");
            src = PageFlowUtil.decode(src);
            Attachment at = _attachments.get(src);
            if (null != at)
            {
                img.setAttribute("src", _attachPrefix + PageFlowUtil.encode(at.getName()));
                wikiDependencies.add(at.getName());
            }
        }

        // back to html
        StringBuilder innerHtml;

        try
        {
            innerHtml = new StringBuilder();

            //look for style elements, as tidy moves them to the head section
            NodeList styleNodes = doc.getElementsByTagName("style");
            for (int idx = 0; idx < styleNodes.getLength(); ++idx)
            {
                innerHtml.append(PageFlowUtil.convertNodeToHtml(styleNodes.item(idx)));
                innerHtml.append('\n');
            }

            Node bodyNode = doc.getElementsByTagName("body").item(0);
            // Uncomment the below to debug bodyNode contents
            // inspectNode(bodyNode, 0);

            String bodyHtml = PageFlowUtil.convertNodeToHtml(bodyNode);

            // Issue 12160: if there were any link decoding exceptions, replace the link with the error message
            if (linkExceptions.size() > 0)
            {
                for (Map.Entry<Element, String> link : linkExceptions.entrySet())
                {
                    String linkHtml = "<span class='labkey-error'>" + link.getValue() + "</span>";
                    bodyHtml = bodyHtml.replace(PageFlowUtil.convertNodeToHtml(link.getKey()), linkHtml);
                }
            }

            innerHtml.append(bodyHtml, "<body>".length(), bodyHtml.length()-"</body>".length());
        }
        catch (Exception e)
        {
            innerHtml = new StringBuilder("<div class=\"labkey-error\"><b>An exception occurred while generating the HTML.  Please correct this content.</b></div><br>The error message was: ");
            innerHtml.append(e.getMessage()); 
            volatilePage = false;
        }

        return new FormattedHtml(innerHtml.toString(), volatilePage, wikiDependencies, Collections.emptySet(), cds);
    }


    // ${labkey.<type>(<any_stream of characters>)}
    // Pattern.DOTALL allows the parameter list to span multiple lines
    private static final Pattern _substitutionPattern = Pattern.compile("\\$\\{labkey\\.(\\w+)\\((.*?)\\)\\}", Pattern.DOTALL);

    // <any word>='<any value>', allowing whitespace before, after, and in-between
    private static final Pattern _paramPattern = Pattern.compile("\\s*(\\w+)\\s*=\\s*'(.*)'\\s*");

    private FormattedHtml handleLabKeySubstitutions(String text)
    {
        if (text == null)
            return new FormattedHtml("");
        
        // Find all substitution templates embedded in wiki text that have the form ${labkey.<type>(<any_stream of characters>)}.
        Matcher webPartMatcher = _substitutionPattern.matcher(text);

        // If we find none, return immediately
        if (!webPartMatcher.find())
            return new FormattedHtml(text);

        List<Definition> definitions = new ArrayList<>(10);
        Map<Definition, List<String>> wikiErrors = new HashMap<>();
        do
        {
            List<String> paramErrors = new ArrayList<>();
            String substitutionType = webPartMatcher.group(1);          // type
            String params = webPartMatcher.group(2).replace(",", "");
            // Parse the parameters with the symbols in parseWith, they can be used in any order
            // as long as they are the same symbol starts and completes a parameter value
            List<String> paramList = new ArrayList<>();
            paramList.add(params);
            String[] parseWith = { "&#39;", "'" };
            for (String parser : parseWith)
            {
                List<String> paramListTemp = new ArrayList<>();
                for (String paramSection : paramList)
                {
                    String[] paramSplit = paramSection.split(parser);
                    for (int i = 0; i < paramSplit.length; i+=2)
                    {
                        paramListTemp.add(paramSplit[i] + (paramSplit.length > i + 1 ? "'" + paramSplit[i+1]  + "'": ""));
                    }
                }
                paramList = paramListTemp;
            }

            Map<String, String> paramMap = new HashMap<>(10);
            for (String param : paramList)
            {
                Matcher paramMatcher = _paramPattern.matcher(param);

                if (paramMatcher.matches())
                {
                    if (paramMap.containsKey(paramMatcher.group(1)))
                        paramErrors.add(param.trim() + ", there are multiple parameters with this name");
                    else
                        paramMap.put(paramMatcher.group(1), paramMatcher.group(2));
                }
                else if (param.trim().length() > 0)
                    paramErrors.add(param.trim());
            }

            // Stick new definition at beginning of list -- we want to replace them in reverse order
            Definition definition = new Definition(substitutionType, webPartMatcher.start(), webPartMatcher.end(), paramMap);
            definitions.add(0, definition);
            wikiErrors.put(definition, paramErrors);
        }
        while(webPartMatcher.find());

        StringBuilder sb = new StringBuilder(text);
        boolean volatilePage = false;
        LinkedHashSet<ClientDependency> cds = new LinkedHashSet<>();

        // Get the corresponding substitution handler for each type and replace template with substitution
        for (Definition definition : definitions)
        {
            SubstitutionHandler handler = _substitutionHandlers.get(definition.getType());
            FormattedHtml substitution;

            if (null != handler)
                substitution = handler.getSubstitution(definition.getParams());
            else
                substitution = new FormattedHtml("<br><font class='error' color='red'>Error: unknown type, \"labkey." + definition.getType() + "\"</font>");

            sb.replace(definition.getStart(), definition.getEnd(), substitution.getHtml());

            if (substitution.isVolatile())
                volatilePage = true;

            cds.addAll(substitution.getClientDependencies());

            List<String> paramErrors = wikiErrors.get(definition);
            if (paramErrors.size() > 0)
            {
                String errorHTML = "<br>";
                for (String error : paramErrors)
                    errorHTML = errorHTML.concat("<font class='error' color='red'>Error with parameter " +
                            error + " in " + definition.getType() + "</font><br><br>");
                sb.insert(definition.getStart() + substitution.getHtml().length(), errorHTML);
            }
        }

        return new FormattedHtml(sb.toString(), volatilePage, cds);
    }


    private static class Definition
    {
        private final String _type;
        private final int _start;
        private final int _end;
        private final Map<String, String> _params;

        private Definition(String type, int start, int end, Map<String, String> params)
        {
            _type = type;
            _start = start;
            _end = end;
            _params = params;
        }

        public String getType()
        {
            return _type;
        }

        private int getEnd()
        {
            return _end;
        }

        private Map<String, String> getParams()
        {
            return _params;
        }

        private int getStart()
        {
            return _start;
        }
    }


    interface SubstitutionHandler
    {
        @NotNull FormattedHtml getSubstitution(Map<String, String> params);
    }
}
