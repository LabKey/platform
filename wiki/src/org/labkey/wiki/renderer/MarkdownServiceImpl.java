/*
 * Copyright (c) 2017-2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.labkey.wiki.renderer;

import org.labkey.api.markdown.MarkdownService;
import org.labkey.api.module.Module;
import org.labkey.api.module.ModuleLoader;
import org.labkey.api.reports.LabkeyScriptEngineManager;
import org.labkey.api.resource.Resource;
import org.labkey.api.services.ServiceRegistry;
import org.labkey.api.util.Path;
import org.labkey.api.util.StringUtilsLabKey;

import javax.script.Invocable;
import javax.script.ScriptEngine;
import javax.script.ScriptException;
import java.io.BufferedInputStream;
import java.io.InputStreamReader;

public class MarkdownServiceImpl implements MarkdownService
{
    private Object mdCompiled;
    private Invocable invocable;

    public MarkdownServiceImpl() throws Exception
    {
        ScriptEngine engine;
        LabkeyScriptEngineManager svc = ServiceRegistry.get().getService(LabkeyScriptEngineManager.class);
        engine = svc.getEngineByName("nashorn");

        Module module = ModuleLoader.getInstance().getModule("Wiki");
        Path path = Path.parse("scripts/").append("markdown-it.js");
        Resource r = module.getModuleResource(path);
        if (r != null && r.isFile() && r.exists())
        {
            engine.eval(new InputStreamReader(new BufferedInputStream(r.getInputStream(), 1024 * 20), StringUtilsLabKey.DEFAULT_CHARSET));
            engine.eval("var md = new markdownit({" +
                    "breaks: true," +
                    "linkify: true" +
                    "})");
            mdCompiled = engine.eval("md");
            invocable = (Invocable) engine;
            invocable.invokeMethod(mdCompiled, "render", "# call render method here to ensure that nashorn compiles this method before use by app");
        }

    }

    @Override
    public String toHtml(String mdText) throws NoSuchMethodException, ScriptException
    {
        // make sure that the source text has the carriage returns escaped and the whole thing encoded
        // otherwise it wont parse right as a js string if it hits a cr or a quote.
        //        mdText = mdText.replace("\n", "\\n");
        //        mdText = PageFlowUtil.filter(mdText);
        //        Object testResult = engine.eval("md.render('" + mdText + "')");

        // Better yet, avoid js injection by using the invokeMethod syntax that takes a java String as a param

        // markdownit wont accept null as input param
        if (null == mdText)
            mdText = "";
        Object html = invocable.invokeMethod(mdCompiled, "render", mdText);
        if (null == html)
            return null;
        // #32468 include selector so we can have markdown-specific styling namespace
        return "<div class=\"lk-markdown-container\">" + html.toString() + "</div>";
    }
}
