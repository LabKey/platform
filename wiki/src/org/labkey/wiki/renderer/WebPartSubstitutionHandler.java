/*
 * Copyright (c) 2008-2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.labkey.wiki.renderer;

import org.apache.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.labkey.api.collections.CaseInsensitiveHashMap;
import org.labkey.api.settings.AppProps;
import org.labkey.api.util.UniqueID;
import org.labkey.api.view.HttpView;
import org.labkey.api.view.NotFoundException;
import org.labkey.api.view.Portal;
import org.labkey.api.view.ViewContext;
import org.labkey.api.view.WebPartFactory;
import org.labkey.api.view.WebPartView;
import org.labkey.api.view.template.ClientDependency;
import org.labkey.api.wiki.FormattedHtml;

import java.io.StringWriter;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Stack;

/**
 * User: adam
 * Date: Jun 27, 2007
 * Time: 2:53:12 AM
 */
public class WebPartSubstitutionHandler implements HtmlRenderer.SubstitutionHandler
{
    private static final Logger LOG = Logger.getLogger(WebPartSubstitutionHandler.class);
    private static final ThreadLocal<Stack<Map>> _paramsStack = ThreadLocal.withInitial(Stack::new);


    @Override
    @NotNull
    public FormattedHtml getSubstitution(Map<String, String> params)
    {
        params = new CaseInsensitiveHashMap<>(params);
        Stack<Map> stack = _paramsStack.get();

        if (stack.contains(params))
            return new FormattedHtml("<br><font class='error' color='red'>Error: recursive rendering</font>");

        stack.push(params);

        try
        {
            String partName = params.get("partName");
            WebPartFactory factory = Portal.getPortalPartCaseInsensitive(partName);

            if (null == factory)
                return new FormattedHtml("<br><font class='error' color='red'>Error: Could not find webpart \"" + partName + "\"</font>");

            String partLocation = params.get("location");

            Portal.WebPart part = new Portal.WebPart();
            part.setName(partName);
            if (partLocation != null)
                part.setLocation(partLocation);
            part.getPropertyMap().putAll(params);

            try
            {
                ViewContext ctx = HttpView.currentContext();

                if (null == ctx)
                    throw new IllegalStateException("ViewContext should be set");

                // Try to ensure that each webpart gets a unique default DataRegionName
                part.setIndex(UniqueID.getRequestScopedUID(ctx.getRequest()));

                WebPartView<?> view = Portal.getWebPartViewSafe(factory, ctx, part);

                if (null == view)
                    throw new NotFoundException("Webpart factory \"" + factory.getName() + "\" did not return a view");

                view.setEmbedded(true);  // Let the webpart know it's being embedded in another page

                String showFrame = params.get("showFrame");

                if (null != showFrame && !Boolean.parseBoolean(showFrame))
                    view.setFrame(WebPartView.FrameType.NONE);

                view.addAllObjects(params);

                //Issue 15609: we need to include client dependencies for embedded webparts
                LinkedHashSet<ClientDependency> dependencies = new LinkedHashSet<>(view.getClientDependencies());

                StringWriter sw = new StringWriter();
                view.include(view, sw);

                return new FormattedHtml(sw.toString(), true, dependencies);  // All webparts are considered volatile... CONSIDER: Be more selective (e.g., query & messages, but not search)
            }
            catch (Throwable e)
            {
                // Let's at least log these exceptions in dev mode
                if (AppProps.getInstance().isDevMode())
                    LOG.error("Error substituting " + partName, e);

                // Return HTML with error
                return new FormattedHtml("<br><font class='error' color='red'>Error substituting " + partName + ": " + e.getMessage() + "</font>");
            }
        }
        finally
        {
            Map m = stack.pop();
            assert m == params : "Stack problem while checking for recursive webpart rendering: popped params didn't match the params that were pushed";
        }
    }
}
