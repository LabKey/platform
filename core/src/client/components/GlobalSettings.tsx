import React, { PureComponent } from 'react';
import { Panel } from 'react-bootstrap';
import { LabelHelpTip } from '@labkey/components';
import FACheckBox from './FACheckBox';
import {GlobalSettingsOptions} from "./models";

const ROW_TEXTS = [
    {
        id: 'SelfRegistration',
        text: 'Allow self sign up',
        tip: 'Users are able to register for accounts when using database authentication. Use caution when enabling this if you have enabled sending email to non-users.',
    },
    {
        id: 'SelfServiceEmailChanges',
        text: 'Allow users to edit their own email addresses',
        tip: 'Users can change their own email address if their password is managed by LabKey Server.',
    },
    {
        id: 'AutoCreateAccounts',
        text: 'Auto-create authenticated users',
        tip: 'Accounts are created automatically when new users authenticate via LDAP or SSO.',
    },
];

interface Props {
    globalSettings: GlobalSettingsOptions;
    canEdit: boolean;
    checkGlobalAuthBox: (id: string) => void;
    authCount: number;
}

export default class GlobalSettings extends PureComponent<Props> {
    render() {
        let rowTexts = ROW_TEXTS;
        const { canEdit, authCount, checkGlobalAuthBox, globalSettings } = this.props;

        // If there are no user-created auth configs, there is no need to show the auto-create users checkbox
        if (authCount == 1) {
            rowTexts = ROW_TEXTS.slice(0, -1);
        }

        const rowTextComponents = rowTexts.map(text => (
            <div className="global-settings__text-row" key={text.id}>
                <FACheckBox
                    key={text.id}
                    checked={globalSettings[text.id]}
                    canEdit={canEdit}
                    onClick={() => checkGlobalAuthBox(text.id)}
                />

                <span className="global-settings__text">
                    {text.text}
                    <LabelHelpTip
                        title="Tip"
                        body={() => {
                            return <div> {text.tip} </div>;
                        }}
                    />
                </span>
            </div>
        ));

        return (
            <Panel>
                <Panel.Heading>
                    <span className="bold-text">Global Settings</span>
                </Panel.Heading>

                <Panel.Body>
                    <div className="bold-text global-settings__title-text"> Sign up and email options</div>

                    {rowTextComponents}
                </Panel.Body>
            </Panel>
        );
    }
}
