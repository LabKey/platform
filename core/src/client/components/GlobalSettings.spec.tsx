import React from 'react';
import { mount, shallow } from 'enzyme';

import renderer from 'react-test-renderer';

import GlobalSettings from './GlobalSettings';
import FACheckBox from './FACheckBox';

describe('<GlobalSettings/>', () => {
    test('Clicking a checkbox toggles the checkbox', () => {
        const checkGlobalAuthBox = jest.fn(() => wrapper.setProps({ SelfRegistration: false }));

        const component =
            <GlobalSettings
                SelfRegistration={true}
                SelfServiceEmailChanges={true}
                AutoCreateAccounts={false}
                canEdit={true}
                authCount={3}
                checkGlobalAuthBox={checkGlobalAuthBox}
            />;
        const wrapper = mount(component);

        expect(wrapper.props()).toHaveProperty('SelfRegistration', true);

        // Click self registration checkbox
        const firstCheckBox = wrapper.find(FACheckBox).first();
        firstCheckBox.simulate('click');
        expect(checkGlobalAuthBox).toHaveBeenCalled();

        expect(wrapper.props()).toHaveProperty('SelfRegistration', false);
    });

    test('An authCount of 1 eliminates the option to auto-create authenticated users', () => {
        const checkGlobalAuthBox = jest.fn(() => wrapper.setProps({ SelfRegistration: false }));
        const component =
            <GlobalSettings
                SelfRegistration={true}
                SelfServiceEmailChanges={true}
                AutoCreateAccounts={false}
                canEdit={true}
                authCount={3}
                checkGlobalAuthBox={checkGlobalAuthBox}
            />;
        const wrapper = mount(component);

        expect(wrapper.find(FACheckBox).length).toBe(3);
        wrapper.setProps({ authCount: 1 });
        expect(wrapper.find(FACheckBox).length).toBe(2);
        expect(wrapper.text()).not.toMatch(/Auto-create authenticated users/);
    });

    test('view-only mode', () => {
        const checkGlobalAuthBox = (id: string): void => {};
        const component =
            <GlobalSettings
                SelfRegistration={true}
                SelfServiceEmailChanges={true}
                AutoCreateAccounts={false}
                canEdit={true}
                authCount={3}
                checkGlobalAuthBox={checkGlobalAuthBox}
            />;

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
