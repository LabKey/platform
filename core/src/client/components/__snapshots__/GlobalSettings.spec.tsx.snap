// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GlobalSettings/> view-only mode 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="panel-heading"
  >
    <span
      className="bold-text"
    >
      Global Settings
    </span>
  </div>
  <div
    className="panel-body"
  >
    <div
      className="bold-text global-settings__title-text"
    >
       Sign up and email options
    </div>
    <div
      className="global-settings__text-row"
    >
      <span
        className="no-highlight clickable "
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-check-square fa-w-14 fa-lg "
          color="#0073BB"
          data-icon="check-square"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M400 480H48c-26.51 0-48-21.49-48-48V80c0-26.51 21.49-48 48-48h352c26.51 0 48 21.49 48 48v352c0 26.51-21.49 48-48 48zm-204.686-98.059l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.248-16.379-6.249-22.628 0L184 302.745l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.25 16.379 6.25 22.628.001z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
      <span
        className="global-settings__text"
      >
        Allow self sign up
        <span
          className="label-help-target"
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-question-circle fa-w-16 fa-1x label-help-icon"
            data-icon="question-circle"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      className="global-settings__text-row"
    >
      <span
        className="no-highlight clickable "
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-check-square fa-w-14 fa-lg "
          color="#0073BB"
          data-icon="check-square"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M400 480H48c-26.51 0-48-21.49-48-48V80c0-26.51 21.49-48 48-48h352c26.51 0 48 21.49 48 48v352c0 26.51-21.49 48-48 48zm-204.686-98.059l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.248-16.379-6.249-22.628 0L184 302.745l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.25 16.379 6.25 22.628.001z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
      <span
        className="global-settings__text"
      >
        Allow users to edit their own email addresses
        <span
          className="label-help-target"
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-question-circle fa-w-16 fa-1x label-help-icon"
            data-icon="question-circle"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      className="global-settings__text-row"
    >
      <span
        className="no-highlight clickable "
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-square fa-w-14 fa-lg "
          color="#adadad"
          data-icon="square"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </span>
      <span
        className="global-settings__text"
      >
        Auto-create authenticated users
        <span
          className="label-help-target"
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-question-circle fa-w-16 fa-1x label-help-icon"
            data-icon="question-circle"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
</div>
`;
